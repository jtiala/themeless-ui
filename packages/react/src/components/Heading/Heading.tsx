import { cn, CommonComponentProps } from "@themeless-ui/utils";
import { createElement, ReactNode } from "react";

type HeadingLevel = 1 | 2 | 3 | 4 | 5 | 6;

export type HeadingProps = {
  /**
   * The level of the heading.
   */
  level: HeadingLevel;

  /**
   * Heading content.
   */
  children?: ReactNode;
} & CommonComponentProps;

const className = cn("heading");

/**
 * Display a section heading.
 */
export function Heading({ level, children, testId }: HeadingProps) {
  const element = "h" + level;

  return createElement(element, { className, "data-testid": testId }, children);
}
