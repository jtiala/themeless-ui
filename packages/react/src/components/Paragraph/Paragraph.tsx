import { CommonComponentProps, cn } from "@themeless-ui/utils";
import { ReactNode } from "react";

export type ParagraphProps = {
  /**
   * Paragraph content.
   */
  children?: ReactNode;
} & CommonComponentProps;

const className = cn("paragraph");

/**
 * A paragraph of text.
 */
export function Paragraph({ children, id, testId }: ParagraphProps) {
  return (
    <p className={className} id={id} data-testid={testId}>
      {children}
    </p>
  );
}
