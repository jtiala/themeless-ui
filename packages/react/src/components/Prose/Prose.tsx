import { CommonComponentProps, cn } from "@themeless-ui/utils";
import { ReactNode } from "react";

export type ProseProps = {
  /**
   * Prose content.
   */
  children?: ReactNode;
} & CommonComponentProps;

const className = cn("prose");

/**
 * A wrapper component for text content. Adds default spacing for the following components:
 * - `<Blockquote>`
 * - `<Heading>`
 * - `<List>`
 * - `<Paragraph>`
 */
export function Prose({ children, id, testId }: ProseProps) {
  return (
    <div className={className} id={id} data-testid={testId}>
      {children}
    </div>
  );
}
