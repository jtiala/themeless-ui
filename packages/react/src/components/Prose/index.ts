export { Prose } from "./Prose";
export type { ProseProps } from "./Prose";
