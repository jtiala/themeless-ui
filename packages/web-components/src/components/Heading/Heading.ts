/* eslint-disable lit/no-invalid-html */
/* eslint-disable lit/binding-positions */
import style from "@themeless-ui/style/dist/heading.css?inline";
import { cn } from "@themeless-ui/utils";
import { unsafeCSS } from "lit";
import { customElement, property } from "lit/decorators.js";
import { html, unsafeStatic } from "lit/static-html.js";
import { TUIComponent } from "../../utils";

type HeadingLevel = 1 | 2 | 3 | 4 | 5 | 6;

const className = cn("heading");

@customElement("tui-heading")
export class Heading extends TUIComponent {
  /**
   * The level of the heading.
   */
  @property({ type: Number })
  level!: HeadingLevel;

  static styles = unsafeCSS(style);

  override render() {
    const tag = unsafeStatic(`h${this.level}`);

    return html`
      <${tag} class="${className}" id="${this.id}" data-testid="${this.testId}">
        <slot></slot>
      </${tag}>
    `;
  }
}

declare global {
  interface HTMLElementTagNameMap {
    "tui-heading": Heading;
  }
}
