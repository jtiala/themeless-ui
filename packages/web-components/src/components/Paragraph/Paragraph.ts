import style from "@themeless-ui/style/dist/paragraph.css?inline";
import { cn } from "@themeless-ui/utils";
import { html, unsafeCSS } from "lit";
import { customElement } from "lit/decorators.js";
import { TUIComponent } from "../../utils";

const className = cn("paragraph");

@customElement("tui-paragraph")
export class Paragraph extends TUIComponent {
  static styles = unsafeCSS(style);

  override render() {
    return html`
      <p class="${className}" id="${this.id}" data-testid="${this.testId}">
        <slot></slot>
      </p>
    `;
  }
}

declare global {
  interface HTMLElementTagNameMap {
    "tui-paragraph": Paragraph;
  }
}
