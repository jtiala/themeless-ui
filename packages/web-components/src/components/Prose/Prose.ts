import style from "@themeless-ui/style/dist/prose.css?inline";
import { cn } from "@themeless-ui/utils";
import { html, nothing, unsafeCSS } from "lit";
import { customElement } from "lit/decorators.js";
import { TUIComponent } from "../TUIComponent";

const className = cn("prose");

@customElement("tui-prose")
export class Prose extends TUIComponent {
  static styles = unsafeCSS(style);

  override render() {
    return html`
      <div
        class="${className}"
        id="${this.id || nothing}"
        data-testid="${this.testId || nothing}"
      >
        <slot></slot>
      </div>
    `;
  }
}

declare global {
  interface HTMLElementTagNameMap {
    "tui-prose": Prose;
  }
}
