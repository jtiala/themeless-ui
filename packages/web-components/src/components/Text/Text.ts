/* eslint-disable lit/no-invalid-html */
/* eslint-disable lit/binding-positions */
import style from "@themeless-ui/style/dist/text.css?inline";
import { cn } from "@themeless-ui/utils";
import { nothing, unsafeCSS } from "lit";
import { customElement, property } from "lit/decorators.js";
import { html, unsafeStatic } from "lit/static-html.js";
import { TUIComponent } from "../TUIComponent";

type TextType =
  | "abbr"
  | "b"
  | "cite"
  | "code"
  | "del"
  | "em"
  | "i"
  | "ins"
  | "kbd"
  | "mark"
  | "q"
  | "s"
  | "samp"
  | "small"
  | "span"
  | "strong"
  | "sub"
  | "sup"
  | "u"
  | "var";

const className = cn("text");

@customElement("tui-text")
export class Text extends TUIComponent {
  /**
   * The type of the rendered element.
   *
   * @default span
   */
  @property()
  type!: TextType;

  /**
   * Only available when `type` is `abbr`.
   *
   * It's recommended to provide a full expansion of the abbreviated term in plain text on the first use of the abbreviation.
   * If it's not viable to provide the expansion as full text, `title` should be used instead.
   */
  @property()
  abbrTitle?: string;

  /**
   * Only available when `type` is `q`, `del`, or `ins`.
   *
   * `q`: A URI of the source of the information quoted.
   * `del` or `ins`: A URI for a resource that explains the change.
   */
  @property()
  cite?: string;

  /**
   * Only available when `type` is `del` or `ins`.
   *
   * Indicates the time and date of the change. Must be a valid date string with an optional time.
   */
  @property()
  dateTime?: string;

  static styles = unsafeCSS(style);

  constructor() {
    super();

    this.type = "span";
  }

  override render() {
    const tag = unsafeStatic(this.type);

    return html`
      <${tag}
        title="${this.abbrTitle || nothing}"
        cite="${this.cite || nothing}"
        dateTime="${this.dateTime || nothing}"
        class="${className}"
        id="${this.id || nothing}"
        data-testid="${this.testId || nothing}"
      >
        <slot></slot>
      </${tag}>
    `;
  }
}

declare global {
  interface HTMLElementTagNameMap {
    "tui-text": Text;
  }
}
