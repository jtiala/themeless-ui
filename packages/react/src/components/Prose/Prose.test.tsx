import {
  componentToJson,
  describe,
  expect,
  it,
  render,
  renderer,
  screen,
} from "../../../test";
import { Heading } from "../Heading";
import { Paragraph } from "../Paragraph";
import { Text } from "../Text";
import { Prose } from "./Prose";

const lipsum =
  "Lorem ipsum dolor sit amet, consectetur adipiscing elit. Donec accumsan ex vel lacinia volutpat. Ut ultricies, mauris et varius finibus, nulla ante efficitur libero, eget convallis odio ex ut neque. Praesent quis odio ac felis pretium pharetra sit amet quis neque. Nam nec libero purus. Maecenas sagittis lobortis lacinia. Aliquam non vehicula metus. Vivamus fringilla ante eget leo pretium cursus. Morbi venenatis diam metus, ut faucibus dolor mollis a. Nulla nec ex sapien. Curabitur a tempor est. Mauris faucibus congue mi, sit amet pulvinar justo porttitor nec. Nunc et eros a est fermentum lacinia id at nisi.";

const defaultProse = (
  <Prose testId="prose">
    <Heading level={1}>Heading</Heading>
    <Paragraph>{lipsum}</Paragraph>
    <Text>Text</Text>
  </Prose>
);

describe("Prose", async () => {
  it("should render the prose", () => {
    expect(componentToJson(renderer.create(defaultProse))).toMatchSnapshot();

    render(defaultProse);

    const prose = screen.getByTestId("prose");

    expect(prose).toBeInTheDocument();
  });
});
