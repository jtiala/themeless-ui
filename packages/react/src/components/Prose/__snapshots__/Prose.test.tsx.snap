// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Prose > should render the prose 1`] = `
<div
  className="tui__prose"
  data-testid="prose"
>
  <h1
    className="tui__heading"
  >
    Heading
  </h1>
  <p
    className="tui__paragraph"
  >
    Lorem ipsum dolor sit amet, consectetur adipiscing elit. Donec accumsan ex vel lacinia volutpat. Ut ultricies, mauris et varius finibus, nulla ante efficitur libero, eget convallis odio ex ut neque. Praesent quis odio ac felis pretium pharetra sit amet quis neque. Nam nec libero purus. Maecenas sagittis lobortis lacinia. Aliquam non vehicula metus. Vivamus fringilla ante eget leo pretium cursus. Morbi venenatis diam metus, ut faucibus dolor mollis a. Nulla nec ex sapien. Curabitur a tempor est. Mauris faucibus congue mi, sit amet pulvinar justo porttitor nec. Nunc et eros a est fermentum lacinia id at nisi.
  </p>
  <span
    className="tui__text"
  >
    Text
  </span>
</div>
`;
